packages/salesforcedx-vscode-apex-replay-debugger/src/breakpoints/checkpointService.ts

```diff
@@ -426,6 +422,166 @@ export class CheckpointService implements TreeDataProvider<BaseNode> {
+              updateError = true;
+              return false;
+            }
+
+            writeToDebuggerOutputWindow(
+              `${localizedProgressMessage}, ${nls.localize(
+                'checkpoint_creation_status_setting_typeref'
+              )}`
+            );
+            progress.report({
+              increment: 50,
+              message: localizedProgressMessage
+            });
+            // For the active checkpoints set the typeRefs using the source/line info
+            if (!setTypeRefsForEnabledCheckpoints()) {
+              updateError = true;
+              return false;
+            }
+
+            writeToDebuggerOutputWindow(
+              `${localizedProgressMessage}, ${nls.localize(
+                'checkpoint_creation_status_clearing_existing_checkpoints'
+              )}`
+            );
+            progress.report({
+              increment: 50,
+              message: localizedProgressMessage
+            });
+            // remove any existing checkpoints on the server
+            const allRemoved: boolean = await checkpointService.clearExistingCheckpoints();
+            if (!allRemoved) {
+              updateError = true;
+              return false;
+            }
+
+            writeToDebuggerOutputWindow(
+              `${localizedProgressMessage}, ${nls.localize(
+                'checkpoint_creation_status_uploading_checkpoints'
+              )}`
+            );
+            progress.report({
+              increment: 70,
+              message: localizedProgressMessage
+            });
+            // This should probably be batched but it makes dealing with errors kind of a pain
+            for (const cpNode of checkpointService.getChildren() as CheckpointNode[]) {
+              if (cpNode.isCheckpointEnabled()) {
+                if (
+                  !(await checkpointService.executeCreateApexExecutionOverlayActionCommand(
+                    cpNode
+                  ))
+                ) {
+                  updateError = true;
+                }
+              }
+            }
+
+            progress.report({
+              increment: 100,
+              message: localizedProgressMessage
+            });
+            writeToDebuggerOutputWindow(
+              `${localizedProgressMessage}, ${nls.localize(
+                'checkpoint_creation_status_processing_complete_success'
+              )}`
+            );
+          }
+        );
+      });
+    } finally {
+      writeToDebuggerOutputWindow(
+        `${nls.localize('long_command_end')} ${localizedProgressMessage}`
+      );
+      let errorMsg = '';
+      if (updateError) {
+        errorMsg = nls.localize(
+          'checkpoint_upload_error_wrap_up_message',
+          nls.localize('sfdx_update_checkpoints_in_org')
+        );
+        writeToDebuggerOutputWindow(errorMsg, true, VSCodeWindowTypeEnum.Error);
+      }
+      telemetryService.sendCheckpointEvent(errorMsg);
+      creatingCheckpoints = false;
```

Comment:
yikes!  `creatingCheckpoints` (defined on line 949) is essentially a global. (meaning, it is used to control logic & flow in several functions, yet aren't local to the functions.  Not of your doing, I know.

Reply:
yeah this whole file needs a hard look for sure.  14 imports, 18 exports.